File: JS/Dynamic Programming/413. Arithmetic Slices.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var numberOfArithmeticSlices = function(nums) {
    const n = nums.length;
    const dp = new Array(n + 1).fill(0);
    for(let i = 2; i <= n; i++) {
        if (nums[i] - nums[i - 1] === nums[i - 1] - nums[i - 2]) {
            dp[i + 1] = dp[i] + 1; 
        }
    }
    return dp.reduce((acc, i) => acc + i, 0);
}

File: Golang/Array/413.Arithmetic-slices.go
package main

func numberOfArithmeticSlices(nums []int) int {
	if len(nums) < 3 {
		return 0
	}
	ret, diff := 0, 0
	for i := 0; i < len(nums)-1; i++ {
		diff = nums[i+1] - nums[i]
		for j := i + 2; j < len(nums); j++ {
			if diff == nums[j]-nums[j-1] {
				ret++
			} else {
				break
			}
		}
	}
	return ret
}
